Readme.md
restiq
======

Smaller, lighter, faster framework for REST APIs.

Lean and fast for low-latency micro-services where overhead is important.
Depending on the app, can serve 20k requests / second or more.

There are not a lot of frills, but provides route mapping, route decoding,
pre-, post- and per-route middleware stacks.  Errors are caught and converted
into HTTP 500 responses.  Unmapped routes return 405 errors.  The calls
themselves can return any HTTP status code.

Optionally some `restify` compatibility.  I was able to swap out restify in a
fairly complex app and have all its unit tests pass, and the app runs 40-50% more
calls per second on restiq than on restify.

    var restiq = require('restiq');
    var app = restiq.createServer();

    app.addRoute('GET', '/', function(req, res, next) {
        res.end('Hello, world.');
        next();
    });
    app.listen(1337);


Objectives
----------

Why yet another framework?  I wanted

- to run as fast the node built-in http.createServer() (or faster; see below)
- user-defined output formats (tbd)
- different output formats call by call (tbd)
- fewer built-ins in favor of more add-ons
- to better understand the components of nodejs web service implementations,
  and there is no better way to learn than by doing


Comparison
----------

A small echo server, parses and returns the url query parameters:

- [restiq] - 20.9k/s
- [http] - 17.6k/s
- [express] - 7.9k/s
- [restify] - 4.6k/s (8k/s using just the http methods)
- [hapi] - 0.2k/s* (1.8k/s with `setNoDelay()`)
  (loop over the hapi sockets hashed in `reply.request.connection._connections`)

\* - there is a res.write() issue with http.ServerResponse.  Calls writing or
     piping the response run at precisely 25 requests/second per connection.
     It is very easily reproducible; the fix is to turn off the Nagle algorithm
     on the response socket with `res.socket.setNoDelay()`.


Overview
--------

A web service responds to requests sent to pathname-like addresses
("routes").  The server extracts the request parameters, looks up the
computation associated with the route, runs it, and returns the generated
response.

Parameters can be embedded in the request path itself (path parameters),
appended to the path in HTTP query string format (a `?` followed by
'&'-separated name=value pairs, eg `?a=1&b=2`), or be in the request body in
HTTP query string format or some other serialization format eg JSON or BSON.
Restiq knows about path params and on-path and in-body HTTP query params.

The computation is composed of a series of steps (the "middleware stack"),
each step a function taking the request, the response thus far, and a callback
to call to indicate that the step is finished, `(req, res, next)`.  The steps
are run in sequence, each called after the preceding one has finished.

The steps are highly configurable.  They can be run on a route-by-route basis
or in common to all routes.  Steps in common can be either before or after the
per-route steps.  In addition, steps can be configured to run after all other
processing is complete even in case of errors.

The restiq request and responses are just node [http.IncomingMessage] and
[http.ServerResponse] objects.

Restiq includes a thin compatibility layer for shimming simple
[restify] applications onto restiq.


Examples
--------

Surprisingly, it is possible to build on top of http and achieve better
throughput than a canonical http server as shown below.  Because RegExps are
very fast in node, extracting path params is only 5% slower.  (Timed with
node-v0.10.29 on an AMD 3.6 GHz 4x Phenom II.)

Canonical server using [http]:

    var http = require('http');
    var querystring = require('querystring');
    var server = http.createServer(function(req, res) {
        req.data = "";
        req.on('data', function(chunk) { req.data += chunk; });
        req.on('end', function() {
            var url = req.url, qs = url.indexOf('?');
            if (qs >= 0) req.params = querystring.parse(url.slice(qs+1));
            res.writeHead(200, {'Content-Type': 'application/json'});
            res.end(JSON.stringify(req.params));
        });
    });
    server.listen(1337, '127.0.0.1');
    // 17.6k/s  wrk -d8s -t2 -c8 'http://localhost:1337/echo?a=1'

With [restiq]:

    var restiq = require('restiq');
    var app = restiq.createServer({readImmediate: 0});
    app.addStep(restiq.mw.parseQueryParams);
    app.addRoute('GET', '/echo', [
        function(req, res, next) {
            res.writeHeader(200, {'Content-Type': 'application/json'});
            res.end(JSON.stringify(req.params));
            next();
        }
    ]);
    app.listen(1337);
    // 20.9k/s  wrk -d8s -t2 -c8 'http://localhost:1337/echo?a=1'

With [restify]:

    var restify = require('restify');
    var app = restify.createServer();
    app.use(restify.queryParser());
    app.get('/echo', function(req, res, next) {
        res.send(200, req.params);
        next();
    });
    app.listen(1337);
    // 4.6k/s  wrk -d8s -t2 -c8 'http://localhost:1337/echo?a=1'

Change just the first two lines to run it under [restiq]:

    var restify = require('restiq');
    var app = restify.createServer({restify: true});
    // ...
    // 16.2k/s  wrk -d8s -t2 -c8 'http://localhost:1337/echo?a=1'


Methods
-------

### restiq( options )
### restiq.createServer( options )

Create a new app.  The `createServer` method returns a newly created app with
no routes and no middleware steps that is not yet listening for connections.
`restiq()` as a function is not a constructor but a builder, it creates a new
app just like `createServer` does.  (Being a builder is similar to `express`,
`createServer` is similar to `http` and `restify`)

The options:

- `debug` - include stack traces in error responses.  Be cautious about
   sending backtraces off-site.
- `setNoDelay` - turn off Nagle write-combining on the current socket.
   This can greatly reduce call latency when responses use write()
   over an internal low-latency network.  Do NOT disable Nagle for
   responses sent over the public internet.
- `restify` - make the response have methods `res.send` for easier
   compatibility with restify.  This eats into the throughput some, so
   use only as needed.
- `createServer` - the function to use to create the server.  It will be
   called with the function(req, res) that processes web requests.
   Exposed for testing, this defaults to `http.createServer`.
- `readBinary` - when reading the request body, gather the chunks into
   a Buffer instead of a utf8 string.  Gathering to string is faster,
   but Buffers are more traditional for binary data.
- `readImmediate` - when reading the request body, the loop can iterate with
   different strategies.  If set to 0 (the default), it uses `setTimeout` which
   supports the highest throughput under load.  Set to 1 for `setImmediate` and
   the highest throughput with just a few active connections.  Set to 2 for
   `on('data')`, which is in between the two -- not as fast as the others, but
   not as slow either.  As a rule of thumb, at 8 active connections or above
   0 (`setTimeout`) will offer the highest throughput.

        var restiq = require('restiq');
        var app = restiq.createServer(options);


### app.listen( port, [hostname], [backlog], [confirmationCallback] )

Start the service.  If given, `confirmationCallback` will be invoked when
the service is ready to receive requests.  Hostname and backlog as for
`http.createServer`.

### app.setErrorHandler( onError(req, res, err, next) )

Use the provided `onError` function to handle middleware errors.  The default error
handler extracts an http status code from `err`, else returns a 500 Internal Server
Error response.

### app.addStep( func, [where] )

Add a processing step to the middleware stack.  Each step is a function
`step(req, res, next)` taking request, response and a next-step callback.
`func` is a step function or an array of step functions. The optional `where`
specifies in which section of the middleware chain to insert the step; the
default is 'use'.

The middleware sections are:

- `setup`, shared steps before the call is routed
- `use`, partially shared steps before the route handlers are run
- (the route handlers, installed with `addRoute`)
- `after`, shared steps after the call successfully finished
- `finally`, shared steps run in every case after all other steps have finished

Middleware steps are run in the above section order, and steps within a
section are run in the order added.  Shared steps are run by all calls.  Use
steps are partially shared, and are run by only those routes that were added
after the use step had already been added.  I.e. if use step and routes are
added interleaved, not all routes will run all use steps; all routes will run
those use steps that are added before the first route is added.  The route
handler steps are defined per route and added with `addRoute()`.

The setup steps provide an opportunity to edit the route, i.e. implement route
aliasing, version mapping, etc.  The use and route steps implement the
call processing proper.  The after steps are for shared post-call wrapup, for
successful calls.  The finally steps are run as the call teardown, and can do
the logging, analytics reporting, etc.

### app.addRoute( method, path, [options], handlers )

Register a path along with a middleware step function (or array of functions)
to handle requests for it.  Returns a route object that can be used to remove
and re-add the route.  Requesting a path that has not been registered or
calling a path with a different GET, POST, etc request method than it was
registered with results in a 405 error.

Paths can embed named parameters, denoted with `/:paramName`.  Named
parameters are extracted and stored into `req.params` (see also
`restiq.mw.parseRouteParams` below).

For restify compatibility, mapped routes execute those `use` steps that
existed when the route was mapped.  In the sequence `use`, `use`, `map(1)`,
`use`, `map(2)`, calls that request route 1 will run only the first two `use`
steps, but calls that request route 2 will run all three.  All calls will run
all `finally` steps (if any).

Options:

- TBD; none right now.


### app.removeRoute( route )

Remove a previously added route.  The removed route can be re-added later with
`addRoute(route)`.


### app.mapRoute( method, path )

For internal use, look up the route for the call.

The mapped route includes the requested `path`, the matching route `name`, the
`tail` of the query string with the query parameters, any named parameter
`vars` included, and the list of `handlers` to run for this request.

For example:

    app.addRoute('GET', '/:color/echo', echoColor)
    app.mapRoute('GET', '/green/echo?a=1&b=2')
    // => {
    //   path: '/green/echo?a=1&b=2',
    //   name: '/:color/echo',
    //   tail: '?a=1&b=2',
    //   vars: {color: "green"},
    //   handlers: [echoColor]
    // }

Note getting the route extracts only the path params; the query string
params can be gotten with `app.mw.parseQueryParams()`.


restiq.mw
---------

A library of pre-written middleware utility functions. Each middleware is
also exposed through a configurable factory function.

### restiq.mw.parseQueryParams( req, res, next )

Merge the query string parameters into `req.params`. `buildParseQueryParams()`
returns this middleware function.

### restiq.mw.parseRouteParams( req, res, next )

Merge the parameters embedded in the request path into `req.params`.  This is
done automatically as soon as the route is mapped, but explicit param parsing
can override these values.  Re-merging allows control of the param source
precedence. `buildParseRouteParams()` returns this middleware function.

### restiq.mw.parseBodyParams( req, res, next )

Merge the query parameters from the body into `req.params`.  Will read
the body with `restiq.mw.readBody` if it has not been read already.
`buildParseBodyParams(options)` returns this middleware function.

Options as for `buildReadBody`.

### restiq.mw.readBody( req, res, next )

Gather up the message that was sent with the http request, and save it in
`req.body`.  This call is safe to call more than once, but sets body only the
first time.

`buildReadBody(options)` returns the readBody middleware function.

Options

- `maxBodySize` - The maximum request body size to allow, in bytes.
Exceeding this value results in a `400 Bad Request` error response.
There is no limit set by default.

### restiq.mw.discardBody( req, res, next )

Reads and discards request body to force the `end` event on the request.
`buildDiscardBody()` returns this middleware function.

### restiq.mw.skipBody( req, res, next )

If the request body is guaranteed to be empty, it is faster to skip waiting
for the `on('end')` event.  Be careful when using this:  if the request has a
body it needs to be consumed. `buildSkipBody` returns this middleware function.


Restify Compatibility Layer
---------------------------

This is what I have so far --


### app.pre( func )

Add shared middleware step to be called before every request, before the
request is routed.  Pre steps are called in the order added.


### app.use( func )

Add shared middleware step to be called before every request after the `pre()`
steps have all finished.  Each routed call will run only those use steps that
existed at the time it was added; use steps added after a route is added will
not be run by that route.  Use steps are run in the order added.


### app.get( path, handler, [handler2, ...] )

Add a GET route, with handlers to run in the order listed


### app.post( path, handler, [handler2, ...] )

Add a POST route, with handlers to run in the order listed


### app.put( path, handler, [handler2, ...] )

Add a PUT route, with handlers to run in the order listed


### app.delete( path, handler, [handler2, ...] )

Add a DELETE route, with handlers to run in the order listed.
This call is also available as `app.del`.


### restiq.queryParser( )

Returns a middleware `function(req, res, next)` that will extract the http query
string parameters and place them in `req.params`


### restiq.authorizationParser( )

Returns a middleware `function(req, res, next)` that will decode an
`Authorization: Basic` header and set the fields `req.authorization.username`,
`req.authorization.basic.username` and `req.authorization.basic.password`.


### restiq.bodyParser( )

Returns a middleware `function(req, res, next)` that will decode the request
body into an object, string or Buffer.  The decoding is ad-hoc based on the
incoming data type, and is not driven by the request headers.


### restiq.acceptParser( )

Sets the response content-encoding to the preferred (first) acceptable
response type specified in the request that is supported by the server.
Restiq assumes the acceptable encodings are listed in order of preference.
Throws a 406 Not Acceptable error if no match is found.


### req.getId( )

returns the request id contained in the request headers.  Unlike restify,
restiq uses a dash `-` if can't find one, and  does not make one up.


### req.version( )

Returns the options.version string that was passed to `createServer()`.


### req.header( name, [defaultValue] )

Return the named header field, or `defaultValue` if that header field was not
specified in the request.


### req.path( )

Returns `req.url`.


### res.header( name, value )

Set a header value, aka `writeHeader`.


### res.get( name )

Read back a set header value.


### res.send( [statusCode], [response] )

Send a response.  The default status code is 200, the default response the
empty string.  The call determines the content type from the response value,
and emits an appropriate header as well.  NOTE:  restify strongly penalizes a
response that does not set the Content-Type header.  Time it yourself.

Turns out restify responses are also extensions of `http.ServerResponse`, so
all the usual write(), writeHead(), end() work as well.


Tips
----

Random observations on building fast REST services

- nodejs `http` has a speed-of-light of around 27k queries per second (empty
  request body, plaintext response)
- having to assemble the body from the chunks limits http to under 24.5k/s
  (that's if not also parsing request params)
- using req.on('data') to assemble the body drops the ceiling to under 19k/s.
  It is much faster to req.read() in an setTimeout loop than to wait for
  events.  Actual times are sensitive to node version, so check.
- query string params are faster to use than REST path params (because routing
  for static paths is a single hash lookup, vs a for loop over a list of
  regexp objects).  Even though path params are faster to extract with a regexp
  than parsing the query string, it does not make up for the routing latency.
- this may be obvious, but passing just path or just query params is faster than
  passing both
- using `res.write()` to reply imposes a throttle of 25 requests per
  connection.  Workaround is to set `res.socket.setNoDelay()` to disable the
  TCP/IP Nagle algorithm.  Only disable for local traffic, never across the internet.


Todo
----

- make sure that buffers are concatinated and returned as buffers

ChangeLog
0.9.2
- use the nodejs built-in querystring for encoding x-www-form-urlencoded responses
- upgrade qhttp for fixed http_parse_query() (decode + into ' ' space)

0.9.1
- fix res.send() to set res._body to raw object, and set res.headers
- fix res.json() to set _body

0.9.0
- process requests async for a large bump in throughput

0.8.1
- fix 0 'use' steps before the mapped route
- make removeRoute accept routes from both mapRoute() and addRoute()

0.8.0
- fix restiq.error.Error* to construct proper Errors
- refactor middleware runner to always call the _finally steps

0.7.6
- restify compat: emit 'uncaughtException' on unhandled mw error
- restify compat: experimental restiq.InvalidCredentialsError

0.7.5
- check that the middleware error.code is a plausible 3 digit http status before using it
- upgrade to qhttp-0.6.0
- document setErrorHandler
- let addStep(func, where) accept its args in either order

0.7.4
- apply restify emulation before the route is mapped, to have res.send available in unit tests

0.7.3
- only set the restify error handler once at start, to allow caller to override
- restify compat: not a restify error if parseBody cannot decode body (with mapParams:false)
- fix parseBody to set to decoded if it decodes to falsy

0.7.2
- add to ci and coverage tests
- test with qnit 0.15.1 with fixed failure exitcode
- fix typo in makeError source string
- use newer qmock
- update dependencies (no need for qmock, yes for querystring)
- guard against possible null deref in rlib

0.7.1
- faster readBody() data chunk concatenation
- list qmock as a dependency (been one since 2015)
- fix: always return a Buffer in readBinary mode
- fix: parseAuthorization mw should run ok without a callback

0.7.0
- middleware builder functions
- maxBodySize readBody and parseReadBodyParams option

0.6.3
- restify compat: provide res._body
- restify compat: suppress errors from `after`
- run tests with qnit
- upgrade aflow to 0.10.1 for speedup

0.6.2
- fix potential null deref on unmapped routes
- new .gitignore
- fix and clean up Readme, thanks @fidian!

0.6.1
- add req.path() compat method
- tolerate double-close
- read request as 'utf8' to not split multi-byte chars

0.6.0
- bump to aflow 0.9.3 (unit tests, track latest)
- bump to qhttp 0.0.6 (track latest)
- fix Restiq.listen() callback to only call once started
- remove qhttp sources
- make Restiq() build apps like express() does

0.5.3
- move http-client into the qhttp package
- change arlib dependency to qhttp

0.5.2
- split http-client out of test-restiq

0.5.1
- speedup: improve readBody readBinary codepath, redo timings

0.5.0
- app.removeRoute method
- [options] param to addRoute (for later)
- refactor app routers into a struct

0.4.3
- bugfix: decodeBody should not change req.body if cannot decode
- compat: make decodeQueryParams populate req.query as well as req.params
- bugfix: propagate errors from parseBody()

0.4.2
- switch to querystring.encode() for now for compatible flat array encodings
- bump to arlib 0.2.4 to pick up http_parse_query fixes
- revert slower array-of-chunks readBinary, toString("binary") converts into utf8 as if latin1

0.4.1
- 20k/s
- default to readImmediate:0
- restify compat default to readImmediate:2, readBinary:true

0.4.0
- acceptParser() restify helper
- fix error classes
- urldecode route params

0.3.0
- deprecate after() and finally()
- tune read loop

0.2.1
- 21.5k/s
- bugfix: write per-instance req.params not global
- speedup: cache route mw stack
- bodyParser(), authParser() restify helpers
- fix encoders, decoders
- bugfix: accepts()
- bugfix: _bodyEof hang at end of call
- bugfix: restify compat error handling
- setNoDelay option
- speedups
- expose createServer
- mw.closeResponse()
- add 'setup' chain
- fix #search query string parsing
- fix 'finally' steps

0.1.0
- improve restify compat (response handling)
- speedup: faster run stack
- fix body param decoding
- fix route matching edge case
- improve error handling
- use() stack

0.0.1
- 17k/s
- basics working


lib/restiq.js
/**
 * rest framework for micro-services
 * ...speed, speed, speed
 *
 * Copyright (C) 2015,2017 Andras Radics
 * Licensed under the Apache License, Version 2.0
 */

'use strict';

var http = require('http');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

// faster to predefine the fields that restiq appends to req
http.IncomingMessage.prototype.headers = null;
http.IncomingMessage.prototype.restiq = null;
http.IncomingMessage.prototype.query = null;
http.IncomingMessage.prototype.params = null;
http.IncomingMessage.prototype.body = "";
http.IncomingMessage.prototype._route = null;
http.IncomingMessage.prototype._bodyEof = undefined;
// restify adds this method, and restify users expect it (eg unit tests)
http.IncomingMessage.prototype.header = function req_header(name) {
    return this.headers[name.toLowerCase()];
};

// also predefine restify compatibility methods added to req and res
http.IncomingMessage.prototype.method = null;
http.IncomingMessage.prototype.accepts = undefined;
http.IncomingMessage.prototype.getId = null;
http.IncomingMessage.prototype.version = null;
//http.IncomingMessage.prototype.headers = {};
//http.IncomingMessage.prototype.statusCode = 200;
http.IncomingMessage.prototype.authorization = null;
http.IncomingMessage.prototype.username = null;
// restify adds this method, and restify users expect it (eg unit tests)
http.ServerResponse.prototype.header = function res_header(name, value) {
    if (value) this.setHeader(name, value);
    else return this.getHeader(name);
}
http.ServerResponse.prototype.send = null;
http.ServerResponse.prototype.get = null;
http.ServerResponse.prototype.json = null;
http.ServerResponse.prototype._body = null;

// accelerate access
http.IncomingMessage.prototype = toStruct(http.IncomingMessage.prototype);
http.ServerResponse.prototype = toStruct(http.ServerResponse.prototype);

var aflow = require('aflow');
var QRoute = require('./qroute');
module.exports = Restiq;
module.exports.mw = require('./rlib');
module.exports.errors = require('./errors');

function Restiq( opts ) {
    if (this === global || !this) return Restiq.createServer(opts);

    this._opts = opts || {};
    this._emulateRestify = this._opts.restify;
    this._setNoDelay = this._opts.setNoDelay;
    this._debug = this._opts.debug;

    // middleware stacks
    this._setup = new Array();                  // pre-routing mw
    this._before = new Array();                 // pre-call mw
    this._routes = {
        GET: new QRoute(),                      // GET routed mw
        POST: new QRoute(),                     // POST mw
        Other: new QRoute(),                    // all other routed mw
    };
    this._after = new Array();                  // post-call mw
    this._finally = new Array();                // teardown mw

    this._server = null;
    this._errorHandler = null;

    if (this._emulateRestify) {
        if (this._opts.readImmediate === undefined) this._opts.readImmediate = 2;
        if (this._opts.readBinary === undefined) this._opts.readBinary = true;
        var restify = require('./qrestify');
        restify.addRestifyErrorHandler(this);
        this._setup.push(restify.emulateRestify);
    }

    // disable listen on new Restiq() to allow using Restiq() as an app builder
    this.listen = function() { throw new Error("new Restiq() cannot listen, use Restiq.createServer()"); }
}
util.inherits(Restiq, EventEmitter);

// bring these closer, for faster access
Restiq.prototype.mw = module.exports.mw;
Restiq.prototype.errors = module.exports.errors;

// restify compat: the list of recognized mime types
// TODO: populate this from the mime types of the configured decoders
Restiq.prototype.acceptable = [
    'application/json',                         // json object
    'text/plain',                               // urlencoded query string
    'application/octet-stream',                 // Buffer object
    'application/javascript',                   // json object ... but as Buffer?
    'application/x-www-form-urlencoded',        // urlencoded query string
];

Restiq.prototype.setErrorHandler = function setErrorHandler( handler ) {
    // handler(req, res, err, next)
    this._errorHandler = handler;
}

Restiq.prototype.setOutputHandler = function setOutputHandler( handler ) {
    // handler( ??? )
    this._outputHandler = handler;
};

Restiq.prototype.setTimeout = function setTimeout( timeout, listener ) {
    this._server.setTimeout(timeout, listener);
};

Restiq.createServer = function createServer( opts ) {
    var app = new Restiq(opts);
    var createServer = http.createServer;
    if (opts) {
        if (opts.createServer) createServer = opts.createServer;
    }

    // undo the listen override
    delete app.listen;

    app._server = createServer( function(req, res, whenDone) {
        var self = app;
        // assign instance vars, else param parsing stores by reference into the parent!
        req.params = {};

        // faster to setNoDelay here than to on-the-fly disable it on write()
        if (self._setNoDelay) res.socket.setNoDelay();

        // run the setup stack before anything else, even routing
        req.restiq = app;
        setImmediate(function() {
            runMiddlewareStack(self, self._setup, req, res, function(err) {
                // error before routing even: clean up and quit
                if (err) return finishMiddlewareStack(self, req, res, err, whenDone);

                // route the call
                var route = self.mapRoute(req.method, req.url);
// TODO: unmapped route should be 404 Not Found, but might be a breaking change
                if (!route) return finishMiddlewareStack(self, req, res, new self.errors.ErrorMethodNotAllowed("route not mapped"), whenDone);
                req._route = route;     // save for parseRouteParams
                if (route._route.type === 'patt') self.mw.parseRouteParams(req, res);

                // build the middleware stack for this call
                var middlewareStack;
                if (route._route.stack) {
                    middlewareStack = route._route.stack;
                }
                else {
                    // restify compat: only run those mw steps that existed when the route was added
                    self._before.limit = route._route.steps;

                    middlewareStack = concatArrays(
                        self._before,
                        route._route.handlers,
                        self._after
                        // self._finally run unconditionally
                    );
                    route._route.stack = middlewareStack;
                }

                // process the request, return the response
                runMiddlewareStack(self, middlewareStack, req, res, function(err) {
                    // _finally steps are run by finishMiddlewareStack
                    // if mocking the server, will be provided with a callback
                    finishMiddlewareStack(self, req, res, err, whenDone);
                });
            });
        });
    });

    return app;
}

// start the service, invoke the callback once started
// This is *not* used to listen to each request arrive
Restiq.prototype.listen = function listen( ) {
    var i, args = new Array();
    for (i=0; i<arguments.length; i++) args.push(arguments[i]);
    var confirmationCallback = (typeof args[args.length - 1] === 'function') ? args.pop() : false;
    args.push(function(err) {
        if (confirmationCallback) confirmationCallback(err);
    });

    this._server.listen.apply(this._server, args);
};

Restiq.prototype.close = function close( cb ) {
    // suppress the error thrown on double-close
    try { this._server.close(cb) }
    catch (err) { if (cb) cb(err); else if (err.message.indexOf('Not running') >= 0) { } else { throw err; } }
};

function concatArrays( /* VARARGS */ ) {
    var i, j, dst = new Array(), nargs = arguments.length;
    for (i=0; i<nargs; i++) {
        var arr = arguments[i], len = (arr.limit >= 0 ? arr.limit : arr.length);
        for (j=0; j<len; j++) dst.push(arr[j]);
    }
    return dst;
}

function runMiddlewareStack( app, middlewareStack, req, res, next ) {
    var i = 0, done = false, len = middlewareStack.length;
    if (len <= 0) return next();
    aflow.repeatUntil(
        function visitor(cb) {
            var handler = middlewareStack[i];
            i++;
            handler(req, res, function(err) {
                if (err === false) err = 'halt mw';
                cb(err, i >= len);
            });
        },
        function whenDone(err) {
            // caller must finish the call
            if (err === 'halt mw') err = null;
            next(err);
        }
    );
}

function finishMiddlewareStack( app, req, res, err, whenDone ) {
    // TODO: what route to emit?
    // TODO: can end up in here before routing, when no route will be set!
    var restifyRoute = req._route ? req._route.name : req.url;

    if (err) app._endWithError(req, res, err);

    // the _finally steps are executed after every call, whether error or not
    runMiddlewareStack(app, app._finally, req, res, function(err2) {

        // fully consume request body, for connection reuse
        if (!req._bodyEof) app.mw.discardBody(req, res);

        if (err || err2) {
            // TODO: mimic the restify route a bit better, this is just a placeholder
            if (app._emulateRestify) try { app.emit('uncaughtException', req, res, restifyRoute, err) } catch (e) { }
            return app._endWithError(req, res, err || err2);
            if (err && err2) console.log("restiq: mw error in 'finally' stack:", err2.stack)
        }
        // restify suppresses errors thrown in `after`
        if (app._emulateRestify) try { app.emit('after', req, res, restifyRoute, err); } catch(e) { }

        // NOTE: do NOT force an end(), a poorly written call can have
        // unfinished queued continuations that may send a response later.
        // Yes, that will mess with call timing / metrics (hooked to 'after').
        // Maybe could set a timeout to force an end after a minute.
        // if (!res.headersSent) res.end();

        if (whenDone) whenDone();
    })
}

Restiq.prototype._endWithError = function _endWithError( req, res, err ) {
    var self = this;

    // be sure to consume all input, even in case of error.
    // Either that, or close the connection and force the client to reconnect.
    if (!req._bodyEof) this.mw.discardBody(req, res);

    // if the response has already been sent, cannot instead send an error
    if (res.headersSent) return;

    // generate an http error response
    if (this._errorHandler) {
        // use the provided errorHandler to generate an http error response
        this._errorHandler(req, res, err, function(err2) {
            if (err2) console.log("restiq: mw error in service errorHandler:", err2.stack);
        });
    }
    else {
        // use built-in logic to extract an http error response from err
        var code = 500, message = "middleware error";
        if (err.statusCode >= 100 && err.statusCode <= 999) { code = err.statusCode, message = err.message; }
        else if (err.code >= 100 && err.code <= 999) { code = err.code, message = err.message; }

        // TODO: have the commit() hook format the response, and
        // TODO: do not set statusCode or emit response here

        // if no response sent yet, send the error
        // TODO: make top-level error response configurable (pluggable handler)
        res.writeHead(code, {'Content-Type': 'text/plain'});
        if (this._debug > 0) message += "; " + err.stack;
        res.end(message);
    }
    return err;
};

// ----------------------------------------------------------------

Restiq.prototype.addStep = function addStep( func, where ) {
    if (typeof func === 'string') { var t = where; where = func; func = t; }
    where = where || 'use';
    if (Array.isArray(func)) {
        for (var i=0; i<func.length; i++) this.addStep(func[i], where);
        return;
    }
    if (typeof func !== 'function') throw new Error("middleware handler must be a function: ", util.inspect(func));
    switch (where) {
    case 'setup':   this._setup.push(func); break;
    case 'use':     this._before.push(func); break;
    case 'after':   this._after.push(func); break;
    case 'finally': this._finally.push(func); break;
    default: throw new Error(where + ": unknown middleware step location");
    }
};

Restiq.prototype.addRoute = function addRoute( method, path, options, handlers ) {
    method = method.toUpperCase();
    if (!handlers) {
        handlers = options;
        options = {};
    }
    if (!Array.isArray(handlers)) handlers = [handlers];
    if (typeof path === 'object') {
        // TODO: restify can set a route with options
        path = path.path;
    }
    var router = this._routes[method];
    if (router) {
        var info = router.addRoute(path, handlers);
    }
    else {
        var routeName = method.toUpperCase() + '::' + path;
        var info = this._routes.Other.addRoute(routeName, handlers);
    }

    // for restify compatibility, only run those mw "use" steps that existed
    // when this route was added.  Any new steps added later will be ignored
    // For express compatibility, run the mw steps added after the route
    // after the route handlers have run.
    info.steps = this._before.length;
    info.method = method;

    // return a route object that can be removed later
    return info;
};

Restiq.prototype.removeRoute = function removeRoute( route ) {
    var router = this._routes[route.method];
    if (router) router.removeRoute(route);
    else this._routes.Other.removeRoute(route);
};

Restiq.prototype.mapRoute = function mapRoute( method, url ) {
    var router = this._routes[method];
    if (router) {
        return router.mapRoute(url);
    }
    else {
        var mappedRoute;
        var routeName = method.toUpperCase() + '::' + url;
        mappedRoute = this._routes.Other.mapRoute(routeName)
        if (mappedRoute) {
            // strip off our prepended extra routing info
            mappedRoute.path = mappedRoute.path.slice(method.length + 2);
            // route.name is almost redundant, but route._route.name has
            // the prepended METHOD:: prefix, while route.name does not
            mappedRoute.name = mappedRoute.name.slice(method.length + 2);
        }
        return mappedRoute;
    }
};

Restiq.prototype = toStruct(Restiq.prototype);

function toStruct( x ) {
    return toStruct.prototype = x;
}
